blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py

```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+    """
+    
+    for image, count in image_counts.items():
+        bin_index = np.digitize(count, bins, right=True) - 1
+        bin_label = bin_labels[bin_index]
+        bin_counts[bin_label].append(image)
 
 def sample_images_for_quality_check(
-    grouped_images_by_date, input_structured_folder, customer_cvt_folder
-):
-    quality_check_images = get_10_random_images_per_date(grouped_images_by_date)
+    grouped_images_by_date: Dict[str, List[str]], 
+    input_structured_folder: str, 
+    customer_quality_check_folder: str, 
+    n_images_to_sample: int
+) -> None:
+    """
+    Samples a specified number of images from each date for quality checking.
+
+    Parameters
+    ----------
+    grouped_images_by_date : Dict[str, List[str]]
+        A dictionary where keys are dates and values are lists of image file names from those dates.
+    input_structured_folder : str
+        The path of the input folder containing images.
+    customer_quality_check_folder : str
+        The destination folder path for the quality check images.
+    n_images_to_sample : int
+        Number of images to sample for each date.
+    """
+    
+    logger.info(f'Sampling {n_images_to_sample} images for quality check.. \n')
+    
+    quality_check_images = get_n_random_images_per_date(grouped_images_by_date, n_images_to_sample)
+    
+    logger.info(f'Quality check images: {quality_check_images} \n')
 
     for key, values in quality_check_images.items():
         for value in values:
             copy_file(
-                "/" + key + "/" + value, input_structured_folder, customer_cvt_folder
+                "/" + key + "/" + value, str(input_structured_folder), str(customer_quality_check_folder)
             )
+            
+def sample_images_for_retraining(
+    sampled_images_by_date: Dict[datetime, List[Tuple[str, datetime, str]]], 
+    input_structured_folder: str, 
+    customer_retraining_folder: str
+) -> None:
+    """
+    Copies sampled images to the specified retraining folder.
+
+    Parameters
+    ----------
+    sampled_images_by_date : Dict[datetime, List[Tuple[str, datetime, str]]]
+        A dictionary mapping dates to lists of image tuples for retraining. Each tuple contains customer name, 
+        upload date, and image name.
+    input_structured_folder : str
+        The path of the input folder containing images.
+    customer_retraining_folder : str
+        The destination folder path for the retraining images.
+
+    Returns
+    -------
+    None
+        The function does not return anything. It performs the operation of copying the sampled images.
+    """
+    
+    for upload_date, images in sampled_images_by_date.items():
+        # Copy the sampled images
+        for image in images:
+            formatted_upload_date = upload_date.strftime("%Y-%m-%d_%H_%M_%S")
+            image_filename = image[2]  # Assuming image is a tuple (customer_name, upload_date, image_name)
```

Comment:
All these assumptions of this data type you're using make the code a bit confusing for a person who tries to read it for the first time without any knowledge.
I would suggest to start creating your own data type to represent an Image, and then we can see how to simplify everything a bit further.

Reply:
This data object could contain the details of the image you're using and for example also the number of detections and stuff like that.

Reply:
I agree! There are a lot of functions and it's difficult to see how they all come together :)

Firstly, I suggest you move all of these functions outside of the `cleaning_pipeline/components/smart_sampling.py` and put them under `cleaning_pipeline/source/sampling_pipeline.py`. We usually have the `components/*.py `to only contain minimal code.

This way the components part remains clean, something like this: 

 ```
 @command_component(name="smart_sampling")
 def smart_sampling(
     input_structured_folder: Input(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_cvt_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_quality_check_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_retraining_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     database_parameters_json: str,
     customer_name: str,
 ):
 
 SmartSampling.sample_images_for_quality_check(...)
 SmartSampling.sample_images_for_retraining(...)
 
 ```

You can create a class for sampling; in the `__init__` you can set attributes that you are now loosely collecting, for example `sample_images_for_retraining`, `n_images_to_sample` and you can also call the rest of the functions in there, whatever you prefer.